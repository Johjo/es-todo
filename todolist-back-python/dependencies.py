from dataclasses import dataclass, field, replace
from enum import Enum
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)



    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        assert (resource_type, resource) in self.factory, f"{resource_type.value} for {resource} must be injected first"
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

