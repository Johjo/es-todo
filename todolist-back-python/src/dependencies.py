from dataclasses import dataclass, field, replace
from enum import Enum
from math import factorial
from typing import Any

class ResourceType(str, Enum):
    use_case = "use_case"
    adapter = "adapter"
    path = "path"
    infrastructure = "infrastructure"
    data = "data"


@dataclass(frozen=True)
class Dependencies:
    factory: dict[Any, Any] = field(default_factory=dict)

    def feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return self._feed(resource_type=resource_type, resource=resource, factory=factory)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.use_case, resource=use_case, factory=use_case_factory)

    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.adapter, resource=port, factory=adapter_factory)

    def feed_path(self, path: str, factory) -> 'Dependencies':
        return self._feed(resource_type=ResourceType.path, resource=path, factory=factory)

    def feed_infrastructure(self, infrastructure: Any, factory):
        return self._feed(resource_type=ResourceType.infrastructure, resource=infrastructure, factory=factory)

    def feed_data(self, data_name: str, value: Any):
        return self._feed(resource_type=ResourceType.data, resource=data_name, factory=lambda _: value)

    def _feed(self, resource_type: ResourceType, resource: Any, factory: Any) -> 'Dependencies':
        return replace(self, factory={**self.factory, (resource_type, resource): factory})

    def get_use_case(self, use_case: Any) -> Any:
        return self._get_resource(resource_type=ResourceType.use_case, resource=use_case)

    def get_query(self, query: Any) -> Any:
        return self.get_use_case(query)

    def get_infrastructure(self, infrastructure) -> Any:
        resource = self._get_resource(resource_type=ResourceType.infrastructure, resource=infrastructure)
        return resource

    def get_adapter(self, adapter: Any) -> Any:
        resource = self._get_resource(resource_type=ResourceType.adapter, resource=adapter)
        return resource

    def get_path(self, path_name: str) -> Any:
        return self._get_resource(resource_type=ResourceType.path, resource=path_name)

    def get_data(self, data_name) -> Any:
        return self._get_resource(resource_type=ResourceType.data, resource=data_name)

    def _get_resource(self, resource_type: ResourceType, resource) -> Any:
        if (resource_type, resource) not in self.factory:
            raise Exception(f"{resource_type.value} for {resource} must be injected first")
        return self.factory[(resource_type, resource)](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()

    def describe(self) -> str:
        def format_item(key):
            (resource_type, resource_name) = key
            resource = self._get_resource(resource_type, resource_name)
            return f"{resource_name}: {resource_type} => {str(type(resource))}"

        return "\n".join([format_item(key) for key in self.factory.keys()])
