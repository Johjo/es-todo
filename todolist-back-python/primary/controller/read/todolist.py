from abc import ABC, abstractmethod
from dataclasses import dataclass

from dependencies import Dependencies
from hexagon.shared.type import TaskKey, TodolistName, TodolistContext, TodolistContextCount, TaskOpen, TaskName


# todo : use a task key instead of an id
@dataclass
class Task:
    id: TaskKey
    name: TaskName
    is_open: TaskOpen

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name
        }


class TodolistSetReadPort(ABC):
    @abstractmethod
    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        pass

    @abstractmethod
    def all_by_name(self) -> list[TodolistName]:
        pass

    @abstractmethod
    def counts_by_context(self, todolist_name: TodolistName) -> list[tuple[TodolistContext, TodolistContextCount]]:
        pass

    @abstractmethod
    def all_tasks(self, todolist_name: TodolistName) -> list[Task]:
        pass


class TodolistReadController:
    def __init__(self, dependencies: Dependencies):
        self.dependencies = dependencies

    def all_todolist_by_name(self) -> list[str]:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.all_by_name()

    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.task_by(todolist_name=todolist_name, task_key=task_key)

    def counts_by_context(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.counts_by_context(todolist_name=todolist_name)

    def export_to_markdown(self):
        return ""

    def to_markdown(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return to_markdown(todolist_set.all_tasks(todolist_name))


def to_markdown(tasks: list[Task]) -> str:
    def task_to_markdown(task: Task) -> str:
        return f"- [{" " if task.is_open else "x"}] {task.name}"
    return "\n".join([task_to_markdown(task) for task in tasks])
