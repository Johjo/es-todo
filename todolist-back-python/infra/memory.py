from uuid import UUID

from expression import Option, Nothing, Some

from hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[str, TodolistSnapshot] = {}

    def by(self, todolist_name: str) -> Option[TodolistSnapshot]:
        if todolist_name not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[todolist_name])

    def save(self, todolist: TodolistSnapshot):
        self.all_todolist[todolist.name] = todolist

    def task_by(self, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[todolist_name].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self) -> list[str]:
        return [name for name in self.all_todolist.keys()]

    def all_tasks(self, todolist_name: str) -> list[TaskSnapshot]:
        if todolist_name not in self.all_todolist:
            return []
        return list(self.all_todolist[todolist_name].tasks)


