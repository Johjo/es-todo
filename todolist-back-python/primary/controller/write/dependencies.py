from abc import ABC
from collections import namedtuple
from dataclasses import dataclass, replace, field
from typing import Any

from hexagon.todolist.aggregate import TodolistSetPort
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.open_task import OpenTask


@dataclass(frozen=True)
class Dependencies(ABC):
    use_case_factory: dict[Any, Any] = field(default_factory=dict)
    adapter_factory: Any = None

    def get_adapter(self, port) -> Any:
        assert self.adapter_factory, f"adapter for {port} must be injected first"
        return self.adapter_factory(self)

    def feed_use_case(self, use_case: Any, use_case_factory: Any) -> 'Dependencies':
        return replace(self, use_case_factory={**self.use_case_factory, use_case: use_case_factory})


    def feed_adapter(self, port: Any, adapter_factory: Any) -> 'Dependencies':
        return replace(self, adapter_factory=adapter_factory)

    def get_use_case(self, use_case: Any) -> Any:
        assert use_case in self.use_case_factory, f"use_case for {use_case} must be injected first"
        return self.use_case_factory[use_case](self)

    @classmethod
    def create_empty(cls) -> 'Dependencies':
        return Dependencies()


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    factories = {
        TodolistCreate: todolist_create_factory,
        OpenTask : open_task_use_case_factory,
    }

    for use_case, factory in factories.items():
        dependencies = dependencies.feed_use_case(use_case, factory)
    return dependencies


def todolist_create_factory(dependencies: Dependencies):
    return TodolistCreate(dependencies.get_adapter(TodolistSetPort))

def open_task_use_case_factory(dependencies: Dependencies):
    return OpenTask(dependencies.get_adapter(TodolistSetPort))
