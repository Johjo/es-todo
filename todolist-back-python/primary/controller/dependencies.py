from functools import reduce
from typing import Any

from dependencies import Dependencies
from hexagon.fvp.read.which_task import WhichTaskQuery
from hexagon.fvp.write.cancel_priority import CancelPriority
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask


def inject_use_cases(dependencies: Dependencies) -> Dependencies:
    use_cases_with_factory: list[Any] = [ResetFvpSession, CancelPriority, ChooseAndIgnoreTaskFvp, TodolistCreate,
                                         OpenTaskUseCase, CloseTask, RewordTask, ImportManyTask, WhichTaskQuery]

    def feed_use_case(dep: Dependencies, use_case: Any) -> Dependencies:
        return dep.feed_use_case(use_case=use_case, use_case_factory=use_case.factory)

    return reduce(feed_use_case, use_cases_with_factory, dependencies)


