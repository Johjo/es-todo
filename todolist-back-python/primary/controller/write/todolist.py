from hexagon.fvp.type import TaskKey
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.todolist.aggregate import TaskSnapshot
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask
from dependencies import Dependencies
from test.secondary.todolist.test_external_todolist_markdown import MarkdownTodolist


class TodolistWriteController:
    def __init__(self, dependencies: Dependencies) -> None:
        self.dependencies = dependencies

    def create_todolist(self, todolist_name: str) -> None:
        use_case: TodolistCreate = self.dependencies.get_use_case(TodolistCreate)
        use_case.execute(todolist_name=todolist_name)

    def open_task(self, todolist_name: str, task_name: str):
        use_case: OpenTaskUseCase = self.dependencies.get_use_case(OpenTaskUseCase)
        use_case.execute(todolist_name=todolist_name, name=task_name)

    def close_task(self, todolist_name: str, task_key: TaskKey):
        use_case: CloseTask = self.dependencies.get_use_case(CloseTask)
        use_case.execute(todolist_name=todolist_name, key=task_key)

    def reword_task(self, todolist_name: str, task_key: TaskKey, new_name: str):
        use_case: RewordTask = self.dependencies.get_use_case(RewordTask)
        use_case.execute(todolist_name, task_key, new_name)

    def import_many_tasks_from_markdown(self, todolist_name: str, markdown: str):
        use_case: ImportManyTask = self.dependencies.get_use_case(ImportManyTask)
        external_todolist = MarkdownTodolist(markdown)
        use_case.execute(todolist_name, external_todolist)

    def choose_and_ignore_task(self, chosen_task: TaskKey, ignored_task: TaskKey):
        use_case : ChooseAndIgnoreTaskFvp = self.dependencies.get_use_case(ChooseAndIgnoreTaskFvp)
        use_case.execute(chosen_task_id=chosen_task, ignored_task_id=ignored_task)

