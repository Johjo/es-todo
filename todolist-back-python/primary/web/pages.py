import urllib
from dataclasses import dataclass
from uuid import UUID

from bottle import template, Bottle, view, response, request, redirect  # type: ignore

from hexagon.fvp.aggregate import NothingToDo, DoTheTask, ChooseTheTask
from hexagon.fvp.read.which_task import TaskFilter
from hexagon.shared.type import TaskKey, TodolistName
from primary.controller.read.final_version_perfected import FinalVersionPerfectedReadController
from primary.controller.read.todolist import TodolistReadController, Task
from dependencies import Dependencies
from primary.controller.write.todolist import TodolistWriteController

bottle_app = Bottle()


@dataclass
class BottleConfig:
    dependencies: Dependencies


bottle_config = BottleConfig(dependencies=Dependencies.create_empty())


@bottle_app.route("/")
@view("index")
def index():
    controller = TodolistReadController(bottle_config.dependencies)
    return {"todolist_name_set": controller.all_todolist_by_name()}


@bottle_app.post("/todo")
def create_todolist():
    todolist_name = get_string_from_request_post("name")
    TodolistWriteController(bottle_config.dependencies).create_todolist(todolist_name)
    redirect(f"/todo/{todolist_name}")


@bottle_app.route("/todo/<todolist_name>/import")
@view("import")
def show_import(todolist_name):
    return {"todolist_name": todolist_name,
            "query_string": "xxxxquery_string",
            "number_of_items": "xxxx_number_of_items",
            "counts_by_context": {"xxxx1": "xxxx2"},
            "urlencode": urllib.parse.quote,
            }


@bottle_app.post("/todo/<todolist_name>/import")
def import_task(todolist_name):
    markdown_import = get_string_from_request_post("markdown_import")
    TodolistWriteController(bottle_config.dependencies).import_many_tasks_from_markdown(todolist_name, markdown_import)
    return redirect(f"/todo/{todolist_name}")


@bottle_app.route("/todo/<todolist_name>")
def show_todolist(todolist_name):
    task_filter = TaskFilter(todolist_name=todolist_name)  # todo mutate this

    which_task = FinalVersionPerfectedReadController(bottle_config.dependencies).which_task(task_filter=task_filter)
    match which_task:
        case NothingToDo():
            return show_todolist_when_no_task(todolist_name=todolist_name)

        case DoTheTask():
            return show_todolist_when_one_task(todolist_name=todolist_name, do_the_task=which_task)

        case ChooseTheTask():
            return show_todolist_when_two_tasks(todolist_name=todolist_name, choose_the_task=which_task)


def show_todolist_when_no_task(todolist_name):
    return template("nothing", {
        "todolist_name": todolist_name,
        "query_string": "xxxx_query_string",
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": {"xxxx1": "xxxx2"},
        "urlencode": urllib.parse.quote,
    })


def show_todolist_when_one_task(todolist_name: str, do_the_task: DoTheTask):
    return template("do_the_task", {
        "task_name": do_the_task.name,
        "task_id": do_the_task.id,
        "todolist_name": todolist_name,
        "query_string": "xxxx_query_string",
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": {"xxxx1": "xxxx2"},
        "urlencode": urllib.parse.quote}, )


def show_todolist_when_two_tasks(todolist_name: str, choose_the_task: ChooseTheTask):
    return template("choose_the_task", {
        "name_1": choose_the_task.name_1,
        "index_1": choose_the_task.id_1,
        "todolist_name": todolist_name,
        "index_2": choose_the_task.id_2,
        "name_2": choose_the_task.name_2,
        "query_string": "xxxx_query_string",
        "number_of_items": "xxxx_number_of_items",
        "counts_by_context": {"xxxx1": "xxxx2"},
        "urlencode": urllib.parse.quote,
    })


@bottle_app.post("/todo/<todolist_name>/item")
def open_task(todolist_name: str):
    task_name = get_string_from_request_post("task_name")
    TodolistWriteController(bottle_config.dependencies).open_task(todolist_name, task_name)
    return redirect(f"/todo/{todolist_name}")


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/close")
def close_task(todolist_name: str, task_key: str):
    TodolistWriteController(bottle_config.dependencies).close_task(todolist_name, task_key=TaskKey(UUID(task_key)))
    return redirect(f"/todo/{todolist_name}")


@bottle_app.post("/todo/<todolist_name>/item/<task_key>/reword")
def reword_task(todolist_name: str, task_key: str):
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.reword_task(todolist_name=todolist_name,
                           task_key=TaskKey(UUID(task_key)),
                           new_name=get_string_from_request_post("new_name"))

    return redirect(f"/todo/{todolist_name}")


@bottle_app.get("/todo/<todolist_name>/item/<task_key>/reword")
def display_reword_task(todolist_name: str, task_key: str):
    task: Task = TodolistReadController(bottle_config.dependencies).task_by(todolist_name=todolist_name,
                                                                            task_key=TaskKey(UUID(task_key)))
    return template("reword", {
        "todolist_name": todolist_name,
        "task_id": task_key,
        "query_string": "xxxx_query_string",
        "task_name": task.name})


@bottle_app.get("/todo/<todolist_name>/export")
@view("export")
def display_export_as_markdown(todolist_name: str):
    return {"markdown_export": TodolistReadController(dependencies=bottle_config.dependencies).to_markdown(todolist_name=todolist_name),
            "todolist_name": todolist_name,
            "query_string": "xxx",
            "number_of_items": "xxx",
            "counts_by_context": {"xxx": "xxx"},
            "urlencode": urllib.parse.quote,
            }

@bottle_app.post('/todo/<todolist_name>/item/choose/<chosen_task>/ignore/<ignored_task>')
def choose_and_ignore_task(todolist_name, chosen_task, ignored_task):
    TodolistWriteController(bottle_config.dependencies).choose_and_ignore_task(chosen_task=chosen_task,
                                                                               ignored_task=ignored_task)
    return redirect(f"/todo/{todolist_name}")

@bottle_app.post('/todo/<todolist_name>/item/<task_key>/cancel_priority')
def cancel_priority(todolist_name, task_key):
    controller = TodolistWriteController(bottle_config.dependencies)
    controller.cancel_priority(task_key=TaskKey(UUID(task_key)))
    return redirect(f"/todo/{todolist_name}")

def get_string_from_request_post(field_name):
    return request.forms.getunicode(field_name)
