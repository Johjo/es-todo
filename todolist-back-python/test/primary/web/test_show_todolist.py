from dataclasses import replace

from approvaltests import verify
from approvaltests.reporters import PythonNativeReporter
from webtest import TestApp

from primary.controller.dependencies import Dependencies
from primary.web.pages import bottle_config
from test.hexagon.todolist.fixture import TodolistFaker, TodolistSetForTest


def test_show_when_no_task(todolist_set: TodolistSetForTest, test_dependencies: Dependencies, app: TestApp,
                           fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies
    todolist_set.feed(replace(fake.a_todolist(), name="my_todolist"))

    response = app.get('/todo/my_todolist')

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())


def test_show_when_one_task(todolist_set: TodolistSetForTest, test_dependencies: Dependencies, app: TestApp,
                            fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies
    todolist_set.feed(
        replace(fake.a_todolist(), name="my_todolist", tasks=[replace(fake.a_task(1), name="buy the milk")]))

    response = app.get('/todo/my_todolist')

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())


def test_show_when_two_tasks(todolist_set: TodolistSetForTest, test_dependencies: Dependencies, app: TestApp,
                            fake: TodolistFaker):
    bottle_config.dependencies = test_dependencies
    todolist_set.feed(
        replace(fake.a_todolist(), name="my_todolist", tasks=[
            replace(fake.a_task(1), name="buy the milk"), replace(fake.a_task(2), name="buy the water")]))

    response = app.get('/todo/my_todolist')

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())
