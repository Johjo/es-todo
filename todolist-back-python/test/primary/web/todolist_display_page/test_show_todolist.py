from datetime import date

from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter  # type: ignore
from webtest import TestApp  # type: ignore

from src.dependencies import Dependencies
from src.infra.memory import Memory
from src.primary.web.pages import bottle_config
from test.fixture import TodolistFaker
from test.primary.web._test_double.calendar_for_test import _CalendarForTest
from test.primary.web.fixture import header_with_good_authentication


def test_show_when_no_task(memory:Memory, calendar: _CalendarForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker) -> None:
    bottle_config.dependencies = test_dependencies
    memory.save(user_key="test@mail.fr", todolist=fake.a_todolist(name="my_todolist").to_snapshot())
    calendar.feed_today(fake.a_date())
    response = app.get('/todo/my_todolist', headers=header_with_good_authentication())

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())


def test_show_when_one_task(memory:Memory, calendar: _CalendarForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker) -> None:
    bottle_config.dependencies = test_dependencies
    calendar.feed_today(fake.a_date())
    memory.save(user_key="test@mail.fr", todolist=fake.a_todolist(name="my_todolist").having(tasks=[fake.a_task(1).having(name="buy the milk")]).to_snapshot())

    response = app.get('/todo/my_todolist', headers=header_with_good_authentication())

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())


def test_show_when_task_has_execution_date(memory:Memory, calendar: _CalendarForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker) -> None:
    bottle_config.dependencies = test_dependencies
    calendar.feed_today(date(2024, 5, 17))
    memory.save(user_key="test@mail.fr", todolist=fake.a_todolist(name="my_todolist").having(
        tasks=[fake.a_task(1).having(name="buy the milk", execution_date=date(2023, 10, 19))]).to_snapshot())

    response = app.get('/todo/my_todolist', headers=header_with_good_authentication())

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())


def test_show_when_two_tasks(memory:Memory, calendar: _CalendarForTest, test_dependencies: Dependencies, app: TestApp, fake: TodolistFaker) -> None:
    bottle_config.dependencies = test_dependencies
    calendar.feed_today(fake.a_date())
    memory.save(user_key="test@mail.fr", todolist=fake.a_todolist().having(
        name="my_todolist",
        tasks=[fake.a_task(1).having(name="buy the milk"), fake.a_task(2).having(name="buy the water")]).to_snapshot())

    response = app.get('/todo/my_todolist', headers=header_with_good_authentication())

    assert response.status == '200 OK'
    verify(str(response.body).replace("\\r\\n", "\r\n"), reporter=PythonNativeReporter())
