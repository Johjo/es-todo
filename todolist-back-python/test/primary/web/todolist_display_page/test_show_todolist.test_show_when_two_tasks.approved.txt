b'<html>

<head>

    <meta charset="UTF-8">

    <title>Todo</title>

</head>

<body>



<p>

    <div>

    <h2> buy the milk </h2> <a href="/todo/my_todolist/item/00000000-0000-0000-0000-000000000001/reword?">Renommer</a>

    </div>

<form action="/todo/my_todolist/item/choose/00000000-0000-0000-0000-000000000001/ignore/00000000-0000-0000-0000-000000000002?" method="post">

    <input type="submit" value="Choisir cette t\xc3\xa2che">

</form>



<form action="/todo/my_todolist/item/00000000-0000-0000-0000-000000000001/cancel_priority?" method="post">

    <input type="submit" value="Cette t\xc3\xa2che n\'est plus prioritaire">

</form>



<form action="/todo/my_todolist/item/00000000-0000-0000-0000-000000000001/close?" method="post">

    <input type="submit" value="C\'est fait">

</form>

</p>



<p>

    <div>

    <h2> buy the water </h2> <a href="/todo/my_todolist/item/00000000-0000-0000-0000-000000000002/reword?">Renommer</a>

    </div>

<form action="/todo/my_todolist/item/choose/00000000-0000-0000-0000-000000000002/ignore/00000000-0000-0000-0000-000000000001?" method="post">

    <input type="submit" value="Choisir cette t\xc3\xa2che">

</form>



<form action="/todo/my_todolist/item/00000000-0000-0000-0000-000000000002/close?" method="post">

    <input type="submit" value="C\'est fait">

</form>

</p>









<div>

Ajouter une t\xc3\xa2che

<form action="/todo/my_todolist/item?" method="post">

    <input type="text" name="task_name">

    <input type="submit" value="Add item">

</form>

</div>



Red\xc3\xa9marrer l\'algo

<form action="/todo/my_todolist/reset?" method="post">

    <input type="submit" value="Reset">

</form>



Il y a xxxx_number_of_items t\xc3\xa2ches en attente.





<h2> Contexte </h2>

<form action="/todo/my_todolist" method="get">

<ul>

<li> <a href="/todo/my_todolist?only_inbox=1">Inbox</a></li>

<li> #context_1 : 5 </li>

    <input type="checkbox" name="include_context" value="#context_1" > Inclure

    <input type="checkbox" name="exclude_context" value="#context_1" > Exclure

<li> #context_2 : 10 </li>

    <input type="checkbox" name="include_context" value="#context_2" > Inclure

    <input type="checkbox" name="exclude_context" value="#context_2" > Exclure

</ul>

    <input type="submit" value="Filtrer">

</form>



<ul>

<li><a href="/todo/my_todolist/export">Exporter les t\xc3\xa2ches</a></li>

<li><a href="/todo/my_todolist/import">Importer des t\xc3\xa2ches</a></li>

</ul>







</body>

</html>'
